File: src/main/java/net/atos/zac/app/informatieobjecten/EnkelvoudigInformatieObjectDownloadService.java
/*
 * SPDX-FileCopyrightText: 2022 Atos
 * SPDX-License-Identifier: EUPL-1.2+
 */

package net.atos.zac.app.informatieobjecten;

import static nl.info.client.zgw.util.ZgwUriUtilsKt.extractUuid;

import java.io.BufferedOutputStream;
import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.net.URI;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.zip.ZipEntry;
import java.util.zip.ZipOutputStream;

import jakarta.inject.Inject;
import jakarta.ws.rs.core.StreamingOutput;

import net.atos.client.zgw.drc.DrcClientService;
import net.atos.client.zgw.zrc.model.ZaakInformatieobject;
import nl.info.client.zgw.drc.model.generated.EnkelvoudigInformatieObject;
import nl.info.client.zgw.zrc.ZrcClientService;

public class EnkelvoudigInformatieObjectDownloadService {
    private static final String RICHTING_INKOMEND = "inkomend";

    private static final String RICHTING_UITGAAND = "uitgaand";

    private static final String RICHTING_INTERN = "intern";

    private static final String SAMENVATTING_BESTANDSNAAM = "samenvatting.txt";

    @Inject
    private DrcClientService drcClientService;

    @Inject
    private ZrcClientService zrcClientService;

    /**
     * Retourneer {@link StreamingOutput} zip-bestand met informatieobjecten en samenvatting
     *
     * @param informatieobjecten lijst van {@link EnkelvoudigInformatieObject} s
     * @return het zip-bestand
     */
    public StreamingOutput getZipStreamOutput(final List<EnkelvoudigInformatieObject> informatieobjecten) {
        return outputStream -> {
            try (final ZipOutputStream zipOutputStream = new ZipOutputStream(new BufferedOutputStream(outputStream))) {
                final Map<String, Map<String, List<String>>> samenvatting = new HashMap<>();
                informatieobjecten.forEach(
                        informatieobject -> samenvattingAddInformatieObject(addInformatieObjectToZip(informatieobject, zipOutputStream),
                                samenvatting));
                zipAddSamenvatting(samenvatting, zipOutputStream);
                zipOutputStream.finish();
            }
            outputStream.flush();
            outputStream.close();
        };
    }

    /**
     * Voeg een informatieobject toe aan het zip-bestand
     *
     * @param informatieobject{@link EnkelvoudigInformatieobject}
     * @param zipOutputStream        {@link ZipOutputStream} van het te updaten zip-bestand
     * @return {@link String} pad naar het toegevoegde bestand in het zip-bestand
     */
    private String addInformatieObjectToZip(
            final EnkelvoudigInformatieObject informatieobject,
            final ZipOutputStream zipOutputStream
    ) {
        final String pad = getInformatieObjectZipPath(informatieobject);
        final ZipEntry zipEntry = new ZipEntry(pad);
        try {
            zipOutputStream.putNextEntry(zipEntry);
            zipOutputStream.write(getInformatieObjectInhoud(extractUuid(informatieobject.getUrl())));
            zipOutputStream.closeEntry();
        } catch (final IOException e) {
            throw new RuntimeException(e);
        }
        return pad;
    }

    /**
     * Retourneer binary inhoud van het informatieobject
     *
     * @param uuid {@link UUID} van het informatieobject
     * @return binary inhoud van het informatieobject
     */
    private byte[] getInformatieObjectInhoud(final UUID uuid) {
        final ByteArrayInputStream inhoud = drcClientService.downloadEnkelvoudigInformatieobject(uuid);
        return inhoud.readAllBytes();
    }

    /**
     * Retourneer pad in mappenstructuur voor een informatieobject in het zip-bestand
     *
     * @param enkelvoudigInformatieobject {@link EnkelvoudigInformatieObject}
     * @return {@link String} pad naar het informatieobject
     */
    private String getInformatieObjectZipPath(final EnkelvoudigInformatieObject enkelvoudigInformatieobject) {
        final List<ZaakInformatieobject> zaakInformatieObjectenList = zrcClientService.listZaakinformatieobjecten(
                enkelvoudigInformatieobject);
        final URI zaakUri = zaakInformatieObjectenList.getFirst().getZaak();
        final String zaakId = zrcClientService.readZaak(zaakUri).getIdentificatie();
        final String subfolder = enkelvoudigInformatieobject.getOntvangstdatum() != null ? RICHTING_INKOMEND :
                enkelvoudigInformatieobject.getVerzenddatum() != null ? RICHTING_UITGAAND :
                RICHTING_INTERN;
        final String[] bestandsnaamExtensie = enkelvoudigInformatieobject.getBestandsnaam().split("\\.");
        return String.format("%s/%s/%s-%s.%s", zaakId, subfolder, bestandsnaamExtensie[0], enkelvoudigInformatieobject.getIdentificatie(),
                bestandsnaamExtensie[1]);
    }

    /**
     * Voeg een informatieobject toe aan de samenvatting
     *
     * @param pad          {@link String} pad naar het informatieobject
     * @param samenvatting {@link Map} samenvatting van het zip-bestand
     */
    private void samenvattingAddInformatieObject(final String pad, final Map<String, Map<String, List<String>>> samenvatting) {
        final String[] padDelen = pad.split("/");
        final String zaakId = padDelen[0];
        final String richting = padDelen[1];
        final String bestandsnaam = padDelen[2];

        // Voeg zaak toe aan samenvatting als deze nog niet bestaat
        if (!samenvatting.containsKey(zaakId)) {
            samenvatting.put(zaakId, new HashMap<>());
        }

        // Voeg richting toe aan samenvatting als deze nog niet bestaat
        final Map<String, List<String>> zaakSamenvatting = samenvatting.get(zaakId);
        if (!zaakSamenvatting.containsKey(richting)) {
            zaakSamenvatting.put(richting, new ArrayList<>());
        }

        // Voeg informatieobject toe
        zaakSamenvatting.get(richting).add(bestandsnaam);
    }

    /**
     * Voeg de samenvatting toe aan het zip-bestand
     *
     * @param samenvatting    {@link Map} samenvatting van bestanden in het zip-bestand
     * @param zipOutputStream {@link ZipOutputStream} van het te updaten zip-bestand
     */
    private void zipAddSamenvatting(final Map<String, Map<String, List<String>>> samenvatting, final ZipOutputStream zipOutputStream) {
        final ZipEntry zipEntry = new ZipEntry(SAMENVATTING_BESTANDSNAAM);
        final StringBuilder stringBuilder = new StringBuilder();

        samenvatting.forEach((zaak, richtingen) -> {
            stringBuilder.append(zaak);
            stringBuilder.append(":\n");
            richtingen.forEach((richting, bestanden) -> {
                stringBuilder.append('\t');
                stringBuilder.append(richting);
                stringBuilder.append(":\n");
                bestanden.forEach(bestand -> {
                    stringBuilder.append("\t  - ");
                    stringBuilder.append(bestand);
                    stringBuilder.append("\n");
                });
            });
            stringBuilder.append('\n');
        });

        try {
            zipOutputStream.putNextEntry(zipEntry);
            zipOutputStream.write(stringBuilder.toString().getBytes());
            zipOutputStream.closeEntry();
        } catch (final IOException e) {
            throw new RuntimeException(e);
        }
    }
}

File: src/main/kotlin/nl/info/zac/app/informatieobjecten/EnkelvoudigInformatieObjectDownloadService.kt
/*
 * SPDX-FileCopyrightText: 2022 Atos, 2026 INFO.nl
 * SPDX-License-Identifier: EUPL-1.2+
 */

package nl.info.zac.app.informatieobjecten

import jakarta.inject.Inject
import jakarta.ws.rs.core.StreamingOutput
import nl.info.client.zgw.drc.DrcClientService
import nl.info.client.zgw.drc.model.generated.EnkelvoudigInformatieObject
import nl.info.client.zgw.util.extractUuid
import nl.info.client.zgw.zrc.ZrcClientService
import nl.info.zac.app.informatieobjecten.exception.EnkelvoudigInformatieObjectDownloadException
import java.io.BufferedOutputStream
import java.io.IOException
import java.util.UUID
import java.util.zip.ZipEntry
import java.util.zip.ZipOutputStream

@Suppress("TooGenericExceptionThrown")
class EnkelvoudigInformatieObjectDownloadService @Inject constructor(
    private val drcClientService: DrcClientService,
    private val zrcClientService: ZrcClientService
) {
    companion object {
        private const val RICHTING_INKOMEND = "inkomend"
        private const val RICHTING_UITGAAND = "uitgaand"
        private const val RICHTING_INTERN = "intern"
        private const val SAMENVATTING_BESTANDSNAAM = "samenvatting.txt"
    }

    fun getZipStreamOutput(informatieobjecten: List<EnkelvoudigInformatieObject>): StreamingOutput =
        StreamingOutput { outputStream ->
            ZipOutputStream(BufferedOutputStream(outputStream)).use { zipOutputStream ->
                val samenvatting = mutableMapOf<String, MutableMap<String, MutableList<String>>>()
                informatieobjecten.forEach { informatieobject ->
                    samenvattingAddInformatieObject(
                        pad = addInformatieObjectToZip(
                            informatieobject = informatieobject,
                            zipOutputStream = zipOutputStream
                        ),
                        samenvatting = samenvatting
                    )
                }
                zipAddSamenvatting(samenvatting = samenvatting, zipOutputStream = zipOutputStream)
                zipOutputStream.finish()
            }
            outputStream.flush()
            outputStream.close()
        }

    private fun addInformatieObjectToZip(informatieobject: EnkelvoudigInformatieObject, zipOutputStream: ZipOutputStream): String {
        val pad = getInformatieObjectZipPath(informatieobject)
        val zipEntry = ZipEntry(pad)
        try {
            zipOutputStream.putNextEntry(zipEntry)
            zipOutputStream.write(getInformatieObjectInhoud(informatieobject.url.extractUuid()))
            zipOutputStream.closeEntry()
        } catch (ioException: IOException) {
            throw EnkelvoudigInformatieObjectDownloadException(
                "Failed to add enkelvoudiginformatieobject with identification '${informatieobject.identificatie}' to zip outputStream",
                ioException
            )
        }
        return pad
    }

    private fun getInformatieObjectInhoud(uuid: UUID): ByteArray =
        drcClientService.downloadEnkelvoudigInformatieobject(uuid).readBytes()

    private fun getInformatieObjectZipPath(enkelvoudigInformatieobject: EnkelvoudigInformatieObject): String {
        val zaakInformatieObjectenList = zrcClientService.listZaakinformatieobjecten(enkelvoudigInformatieobject)
        val zaakUri = zaakInformatieObjectenList.first().zaak
        val zaakId = zrcClientService.readZaak(zaakUri).identificatie
        val subfolder = when {
            enkelvoudigInformatieobject.ontvangstdatum != null -> RICHTING_INKOMEND
            enkelvoudigInformatieobject.verzenddatum != null -> RICHTING_UITGAAND
            else -> RICHTING_INTERN
        }
        val bestandsnaam = enkelvoudigInformatieobject.bestandsnaam
        val baseName = bestandsnaam.substringBeforeLast(".")
        val extension = bestandsnaam.substringAfterLast(".", "")

        val naamMetIdentificatie = when {
            baseName.isNotEmpty() -> "$baseName-${enkelvoudigInformatieobject.identificatie}"
            else -> "-${enkelvoudigInformatieobject.identificatie}"
        }

        return if (extension.isNotEmpty()) {
            "$zaakId/$subfolder/$naamMetIdentificatie.$extension"
        } else {
            "$zaakId/$subfolder/$naamMetIdentificatie"
        }
    }

    private fun samenvattingAddInformatieObject(
        pad: String,
        samenvatting: MutableMap<String, MutableMap<String, MutableList<String>>>
    ) {
        val (zaakId, richting, bestandsnaam) = pad.split("/")
        samenvatting.getOrPut(zaakId) { mutableMapOf() }
            .getOrPut(richting) { mutableListOf() }
            .add(bestandsnaam)
    }

    private fun zipAddSamenvatting(samenvatting: Map<String, Map<String, List<String>>>, zipOutputStream: ZipOutputStream) {
        val zipEntry = ZipEntry(SAMENVATTING_BESTANDSNAAM)
        val stringBuilder = StringBuilder()

        samenvatting.forEach { (zaak, richtingen) ->
            stringBuilder.append(zaak)
            stringBuilder.append(":\n")
            richtingen.forEach { (richting, bestanden) ->
                stringBuilder.append('\t')
                stringBuilder.append(richting)
                stringBuilder.append(":\n")
                bestanden.forEach { bestand ->
                    stringBuilder.append("\t  - ")
                    stringBuilder.append(bestand)
                    stringBuilder.append("\n")
                }
            }
            stringBuilder.append('\n')
        }

        try {
            zipOutputStream.putNextEntry(zipEntry)
            zipOutputStream.write(stringBuilder.toString().toByteArray())
            zipOutputStream.closeEntry()
        } catch (ioException: IOException) {
            throw RuntimeException(ioException)
        }
    }
}
